contracts/buy-and-make/RevenueBuyBack.sol

```diff
@@ -0,0 +1,288 @@
+        }
+    }
+
+    /***************************************
+                    EXTERNAL
+    ****************************************/
+
+    /**
+     * @dev Simply transfers the mAsset from the sender to here
+     * @param _mAsset Address of mAsset
+     * @param _amount Units of mAsset collected
+     */
+    function notifyRedistributionAmount(address _mAsset, uint256 _amount) external override {
+        // Transfer from sender to here
+        IERC20(_mAsset).safeTransferFrom(msg.sender, address(this), _amount);
+
+        emit RevenueReceived(_mAsset, _amount);
+    }
+
+    /**
+     * @notice Buys reward tokens, eg MTA, using mAssets like mUSD or mBTC from protocol revenue.
+     * @param _mAssets Addresses of mAssets that are to be sold for rewards. eg mUSD and mBTC.
+     */
+    function buyBackRewards(address[] calldata _mAssets) external keeperOrGovernor {
+        uint256 len = _mAssets.length;
+        require(len > 0, "Invalid args");
+
+        // for each mAsset
+        for (uint256 i = 0; i < len; i++) {
+            // Get config for mAsset
+            RevenueBuyBackConfig memory config = massetConfig[_mAssets[i]];
+            require(config.bAsset != address(0), "Invalid mAsset");
+
+            // STEP 1 - Redeem mAssets for bAssets
+            IMasset mAsset = IMasset(_mAssets[i]);
+            uint256 mAssetBal = IERC20((_mAssets[i])).balanceOf(address(this));
+            uint256 minBassetOutput = (mAssetBal * config.minBassetPercentage) /
+                BASSET_PECENTAGE_SCALE;
+            uint256 bAssetAmount = mAsset.redeem(
+                config.bAsset,
+                mAssetBal,
+                minBassetOutput,
+                address(this)
+            );
+
+            // STEP 2 - Swap bAssets for rewards using Uniswap V3
+            uint256 minRewardsAmount = (bAssetAmount * config.minBasset2RewardsPrice) /
+                REWARDS_PRICE_SCALE;
+            IUniswapV3SwapRouter.ExactInputParams memory param = IUniswapV3SwapRouter
+            .ExactInputParams(
+                config.uniswapPath,
+                address(this),
+                block.timestamp,
+                bAssetAmount,
+                minRewardsAmount
+            );
+            uniswapRouter.exactInput(param);
+        }
+    }
+
+    /**
+     * @notice donates purchased rewards, eg MTA, to staking contracts via the Emissions Controller.
+     */
+    function donateRewards() external keeperOrGovernor {
+        // STEP 1 - Get the voting power of the staking contracts
+        uint256 numberStakingContracts = stakingDialIds.length;
+        uint256[] memory votingPower = new uint256[](numberStakingContracts);
+        uint256 totalVotingPower;
+        // Get the voting power of each staking contract
+        for (uint256 i = 0; i < numberStakingContracts; i++) {
+            address staingContractAddress = EMISSIONS_CONTROLLER.stakingContracts(stakingDialIds[i]);
```

Comment:
Looking up stakingContracts with dialId is incorrect. Position in dials array != position in stakingContracts array. You can just look up the dialData.address to get the address of the staking contract.

Reply:
Good spot. Thanks